factory-containers/compose_app_downloader.py

```diff
@@ -0,0 +1,217 @@
+        self._process.wait(timeout=60)
+
+        shutil.rmtree(self.docker_dir)
+        self._remove_non_image_data()
+        logger.info("Docker daemon has been stopped")
+
+    def _remove_non_image_data(self):
+        data_subdirs = os.listdir(self.data_root)
+        for dir in data_subdirs:
+            if dir not in self.ImageDataDirs:
+                shutil.rmtree(os.path.join(self.data_root, dir), ignore_errors=True)
+
+
+class ImageDownloader:
+    def __init__(self, env):
+        self._env = os.environ.copy()
+        self._env.update(env)
+
+    def pull(self, platform, url):
+        cmd = self._get_cmd(platform, url)
+        logger.info('Fetching image: {}'.format(cmd))
+        subprocess.run(cmd.split(), check=True, stdout=None, stderr=None, env=self._env)
+
+    def _get_cmd(self, platform: str, url: str):
+        pass
+
+    def download_images(self, app: ComposeApp, platform, dst_dir: str):
+        for service_name, service_cfg in app.services():
+            logger.info('Copying image {} to {}'.format(service_cfg['image'], dst_dir))
+            self.pull(platform, service_cfg['image'])
+
+
+class DockerDownloader(ImageDownloader):
+    def __init__(self, dst_daemon_host):
+        self.dst_daemon_host = dst_daemon_host
+
+        super().__init__({'DOCKER_HOST': self.dst_daemon_host})
+
+    def _get_cmd(self, platform: str, url: str):
+        return 'docker pull --platform {} {}'.format(platform, url)
+
+
+class SkopeoDownloader(ImageDownloader):
+    def __init__(self, dst_daemon_host):
+        self._dst_daemon_host = dst_daemon_host
+
+        super().__init__({'DOCKER_HOST': self._dst_daemon_host})
+
+    def _get_cmd(self, platform, url: str):
+        if -1 == url.find('@'):
+            url_parts = url.split(':')
+            sha = url_parts[1]
+        else:
+            url_parts = url.split('@')
+            sha = url_parts[1].split(':')[1][:9]
+
+        return 'skopeo --override-arch {} --override-os linux' \
+               ' copy --dest-daemon-host {} docker://{} docker-daemon:{}:{}'.format(
+            platform, self._dst_daemon_host, url, url_parts[0], sha)
+
+
+# download app's images and store them in dockerd:overlay2 format
+# they will NOT be stored in OCI or docker image format or flatten image rootfs/bundle
+# the resultant file tree structure can be consumed solely by dockerd:overlay2 graph driver
+@contextlib.contextmanager
+def ComposeAppDownloader(arch, dst_dir, image_downloader_cls: ImageDownloader):
+    with ContainerDaemon() as containerd:
+        with DockerDaemon(dst_dir, containerd.address) as dockerd:
+            downloader = image_downloader_cls(dockerd.host)
+
+            def download_app(app: ComposeApp):
+                for service_name, service_cfg in app.services():
+                    logger.info('Copying image {} to {}'.format(service_cfg['image'], dst_dir))
+                    downloader.pull(arch, service_cfg['image'])
+
+            yield download_app
```

Comment:
slick use of chaining together some context managers.

Reply:
@doanac https://github.com/foundriesio/meta-lmp/pull/195 has been merged, so I suppose we should merge this PR too.